Refactor: Let ExoPlayer handle audio focus

jakobkukla opened the pull request:
ExoPlayer can handle audio focus, which makes pretty much all the manual handling obsolete. One important thing to note is that this would remove the setting to disable audio ducking, because ExoPlayer always lowers volume on incoming notifications or navigation tts. This also fixed #62.

dkanada commented:
The code changes look great, I'm wondering if the feature itself should be removed though. The current implementation would technically work when Chromecast and Session playback are added, which might be desirable? Not quite sure to be honest, this is definitely the simpler solution.

jakobkukla commented:
Well I'm not really sure if this would be needed in a chromecast implementation. I believe Cast Application Framework manages audio focus automatically. (I'm not sure though haha). And also there is a point to make that eg audio ducking is not really a thing if you're casting. I think the logic would only grew more complex over time and would probably be split up anyways in the long run.

What's Session playback?

dkanada commented:
Session playback is the feature in Jellyfin that no one seems to know about where you can remote control any other Jellyfin client pretty easily. It's a bit buggy on web right now to be honest, but the vue project is already implementing it a lot better so I'm super excited to see it improve. I'd also like to add it to Gelli at some point, hence why I mentioned it.

I'm inclined to agree with you - audio focus probably isn't relevant for session or chromecast playback. One last reservation I have with the change is that right now, audio playback in Gelli is very "push" oriented, and there are a [number of events](https://github.com/dkanada/gelli/blob/master/app/src/main/java/com/dkanada/gramophone/service/playback/Playback.java#L24-L32) that don't get sent back to the audio service automatically. Do the widgets, notification, and interface all still update properly when the audio is paused after this change? Just want to make sure. In the future I need to move those four methods into the callback below, but it's an involved change and I haven't had the time yet.

dkanada commented on app/src/main/java/com/dkanada/gramophone/service/playback/LocalPlayer.java:
```diff
@@ -82,7 +84,12 @@ public LocalPlayer(Context context) {
+                .setContentType(C.CONTENT_TYPE_MUSIC)
+                .build();
+
+        exoPlayer = new SimpleExoPlayer.Builder(context).setMediaSourceFactory(mediaSourceFactory).setAudioAttributes(audioAttributes, true).build();
```
```suggestion
        exoPlayer = new SimpleExoPlayer.Builder(context)
            .setAudioAttributes(audioAttributes, true)
            .setMediaSourceFactory(mediaSourceFactory)
            .build();
```

dkanada commented on app/src/main/java/com/dkanada/gramophone/service/playback/LocalPlayer.java:
```diff
@@ -82,7 +84,12 @@ public LocalPlayer(Context context) {
         MediaSourceFactory mediaSourceFactory = new UnknownMediaSourceFactory(buildDataSourceFactory());
-        exoPlayer = new SimpleExoPlayer.Builder(context).setMediaSourceFactory(mediaSourceFactory).build();
+        AudioAttributes audioAttributes = new AudioAttributes.Builder()
+                .setUsage(C.USAGE_MEDIA)
+                .setContentType(C.CONTENT_TYPE_MUSIC)
+                .build();
```
```suggestion
            .setUsage(C.USAGE_MEDIA)
            .setContentType(C.CONTENT_TYPE_MUSIC)
            .build();
```

jakobkukla commented:
Ok, so I just rebased and tested it again. I was running my build the last few days and couldn't find any bugs/inconsistencies with the UI. About the getVolume() and isPlaying() methods there is this to keep in mind (from the official [exoplayer dev blog](https://medium.com/google-exoplayer/easy-audio-focus-with-exoplayer-a2dcbbe4640e)):

>**getPlayWhenReady**
>The value returned by getPlayWhenReady may surprise you in one case:
>
>When playback temporarily pauses in response to a transient loss of audio focus, getPlayWhenReady() continues to return true. The reasoning is that the content will continue playing when audio focus is regained.
>
>**getVolume**
>When the volume is temporarily ducked due to a transient loss of audio focus, getVoume() returns the last value that you set (with setVolume()) — not the volume that you hear. This means setVolume and getVolume always agree. What you set is what you get… but not necessarily what’s heard.

But I haven't noticed any problems with this change in behavior.

On another node: Did you suggest any changes to the code? I can't see them on GitHub but I got two mail notifications with requested changes. In either way I fixed the few formatting issues.

jakobkukla commented:
@dkanada fixed merge conflicts!
